Jetsnack/benchmark/src/main/java/com/example/jetsnack/benchmark/StartupBenchmark.kt

```diff
@@ -0,0 +1,103 @@
+import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+/**
+ * Run this benchmark from Studio to see startup measurements, and captured system traces
+ * for investigating your app's performance from a cold state.
+ */
+@RunWith(AndroidJUnit4ClassRunner::class)
+class ColdStartupBenchmark : AbstractStartupBenchmark(StartupMode.COLD)
+
+/**
+ * Run this benchmark from Studio to see startup measurements, and captured system traces
+ * for investigating your app's performance from a warm state.
+ */
+@RunWith(AndroidJUnit4ClassRunner::class)
+class WarmStartupBenchmark : AbstractStartupBenchmark(StartupMode.WARM)
+
+/**
+ * Run this benchmark from Studio to see startup measurements, and captured system traces
+ * for investigating your app's performance from a hot state.
+ */
+@RunWith(AndroidJUnit4ClassRunner::class)
+class HotStartupBenchmark : AbstractStartupBenchmark(StartupMode.HOT)
+
+/**
+ * Base class for benchmarks with different startup modes.
+ * Enables app startups from various states of baseline profile or [CompilationMode]s.
+ */
+abstract class AbstractStartupBenchmark(private val startupMode: StartupMode) {
+    @get:Rule
+    val benchmarkRule = MacrobenchmarkRule()
+
+    /**
+     * Tests the worst case startup performance. No baseline profile, no pre-compilation.
+     */
+    @Test
+    fun startupNoCompilation() = startup(CompilationMode.None())
+
+    /**
+     * Tests the average startup time a user will see if we didn't include a baseline profile.
+     * This is the test you can compare your baseline profile time to to make sure you are actually seeing
+     * a benefit.
+     */
+    @Test
+    fun startupBaselineProfileDisabled() = startup(
+        CompilationMode.Partial(
+            baselineProfileMode = BaselineProfileMode.Disable,
+            warmupIterations = 1
+        )
+    )
+
+    /**
+     * Tests the average startup time a user will see with our baseline profile. This is the test
+     * you are trying to optimise when working out what to include in your profile. If everything
+     * is working correctly this test should have the best performance except for the full
+     * compilation.
+     */
+    @Test
+    fun startupBaselineProfile() =
+        startup(CompilationMode.Partial(baselineProfileMode = BaselineProfileMode.Require))
+
+    /**
+     * Tests the theoretical best case scenario if everything was pre-compiled.
+     */
+    @Test
+    fun startupFullCompilation() = startup(CompilationMode.Full())
+
+    private fun startup(compilationMode: CompilationMode) = benchmarkRule.measureRepeated(
+        packageName = "com.example.jetsnack",
+        metrics = listOf(StartupTimingMetric()),
+        compilationMode = compilationMode,
+        iterations = 10,
+        startupMode = startupMode
+    ) {
+        pressHome()
+        startActivityAndWait()
```

Comment:
`pressHome()` should be a part of the `setup` block.

This is why your benchmark numbers are a bit unstable.

Also, as a bonus consider call `reportFullyDrawn()` from inside your `ComposableActivity` in a follow up PR  (when the top level composable is ready) so you can actually see the complete costs.

Reply:
Thanks for the review Rahul. Super helpful.

Reply:
Good catch, Rahul!
This is copied from the BaselineProfileRule's  [documentation](https://developer.android.com/reference/androidx/benchmark/macro/junit4/MacrobenchmarkRule) code snippet, so needs fixing there as well. I filed a [bug on this](https://issuetracker.google.com/223352609).

Reply:
Why does it actually make the results a bit unstable? 🤔 
Also, are there any other calls which can make the results unstable? Idk `device.waitForIdle`? 
Or generally should we just `pressHome()` if `StartupMode.HOT`? Otherwise I don't think it's necessary, because the activity/process is killed anyway?

Reply:
`pressHome()` manipulates the task stack, and makes the launcher visible (which is a different app). This might cause interference because the launcher might be in the middle of something, and thus `pressHome()` is not doing something deterministic.

`device.waitForIdle()` on the other hand is just waiting for the main looper in the app to be idle. So it's a safer thing to do (because the only thing that can introduce non-deterministic behavior is the app itself).